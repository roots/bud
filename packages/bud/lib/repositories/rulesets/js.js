"use strict";
exports.__esModule = true;
exports.js = void 0;
var js = function (bud) {
    return bud.hooks.filter('webpack.module.rules.js', {
        test: bud.hooks.filter('webpack.module.rules.js.test', bud.patterns.get('js')),
        exclude: bud.hooks.filter('webpack.module.rules.js.exclude', bud.patterns.get('vendor')),
        use: bud.hooks.filter('webpack.module.rules.js.use', {
            value: [bud.uses.get('babel')(bud)],
            bud: bud
        }).value
    });
};
exports.js = js;
//# sourceMappingURL=js.js.map