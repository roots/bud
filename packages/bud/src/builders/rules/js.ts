import type {Rule} from '@roots/bud-typings'

export const test: Rule.Factory<Rule.Conditional> = ({
  patterns,
}) => patterns.get('js')

export const exclude: Rule.Factory<Rule.Conditional> = ({
  patterns,
}) => patterns.get('modules')

export const use: Rule.Factory<Rule.Use> = ({build}) => [
  build.getItem('thread'),
  build.getItem('cache'),
  build.getItem('raw'),
]
