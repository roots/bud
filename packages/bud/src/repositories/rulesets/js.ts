import type {Bud} from '../..'
import type {WebpackRule} from '@roots/bud-typings'

const js = (bud: Bud): WebpackRule =>
  bud.hooks.filter('webpack.module.rules.js', {
    test: bud.hooks.filter(
      'webpack.module.rules.js.test',
      bud.patterns.get('js'),
    ),

    exclude: bud.hooks.filter(
      'webpack.module.rules.js.exclude',
      bud.patterns.get('vendor'),
    ),

    use: bud.hooks.filter('webpack.module.rules.js.use', {
      value: [bud.uses.get('babel')(bud)],
      bud,
    }).value,
  })

export {js}
