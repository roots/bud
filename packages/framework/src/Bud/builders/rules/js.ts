import type {Rule} from '@roots/bud-typings'

export const test: Rule.Factory<Rule.Conditional> = bud =>
  bud.patterns.get('js')

export const exclude: Rule.Factory<Rule.Conditional> = bud =>
  bud.patterns.get('modules')

export const use: Rule.Factory<Rule.Use> = bud => [
  bud.build.items.raw.make(),
]
