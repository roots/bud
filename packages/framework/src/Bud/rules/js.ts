export const test: Framework.Rule.Factory<Framework.Rule.Conditional> = bud =>
  bud.patterns.get('js')

export const exclude: Framework.Rule.Factory<Framework.Rule.Conditional> = bud =>
  bud.patterns.get('modules')

export const use: Framework.Rule.Factory<Framework.Rule.Use> = bud => [
  bud.build.items.raw.make(),
]
