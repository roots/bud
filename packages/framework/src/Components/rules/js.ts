export const test: Build.Rule.Factory<Build.Rule.Conditional> = bud =>
  bud.store['patterns'].get('js')

export const exclude: Build.Rule.Factory<Build.Rule.Conditional> = bud =>
  bud.store['patterns'].get('modules')

export const use: Build.Rule.Factory<Build.Rule.Use> = bud => [
  bud.components['items'].get('babel-loader').make(),
]
