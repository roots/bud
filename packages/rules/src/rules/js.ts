import Bud from '@roots/bud-types'

export const test: Bud.Rule.Factory<Bud.Rule.Conditional> = function () {
  return this.bud.patterns.get('js')
}

export const exclude: Bud.Rule.Factory<Bud.Rule.Conditional> = function () {
  return this.bud.patterns.get('modules')
}

export const use: Bud.Rule.Factory<Bud.Rule.Loader> = function () {
  return this.bud.loaders.get('babel').webpack
}
