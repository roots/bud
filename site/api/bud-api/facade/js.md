---
id: bud-api.facade.js
title: js property
sidebar_label: js property
hide_title: true
sidebar: "api"
slug: js
---

## Facade.js property

Generate application entrypoints from source asset paths.

Signature:

```typescript
js: entryFacade;
```

## Remarks

\*\*Globbing\*\*

Uses \[fast-glob\](https://git.io/JkGbw) syntax.

\*\*Supported patterns\*\*

- `*` matches any number of characters, but not `/` - `?` matches a single character, but not `/` - `**` matches any number of characters, including `/`, as long as it's the only thing in a path part - `{}` allows for a comma-separated list of "or" expressions - `!` at the beginning of a pattern will negate the match

## Example 1

Create an entrypoint from a single file:

```js
app.entry("app", "app.js");
```

## Example 2

Create an entrypoint from multiple files:

```js
app.entry("app", ["js/app.js", "css/app.css"]);
```

## Example 3

Create an entrypoint comprised of all js assets:

```js
app.entry("app", "*.js");
```

## Example 4

You may create more than one entrypoint using object syntax:

```js
app.entry({
  scripts: "*.js",
  styles: ["*.css", "*.scss"],
});
```

## Example 5

Declare entrypoint dependencies:

```js
app.entry({
  react: {
    import: ["react", "react-dom"],
  },
  app: {
    import: ["app.js"],
    dependOn: ["react"],
  },
});
```
