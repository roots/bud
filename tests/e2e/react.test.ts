import {afterAll, beforeAll, describe, expect, it} from 'vitest'

import {close, page, path, read, setup, update} from './runner/index.js'

describe(`html output of examples/react`, () => {
  let original: string | undefined

  beforeAll(async () => {
    await setup(`react`).then(({page}) => page)
    original = await read(`src`, `components`, `App.js`)
  })

  afterAll(close)

  it(`should have expected default state`, async () => {
    expect(original).toMatchSnapshot()

    expect(
      await page
        .$(`.header img`)
        .then(async handle => await handle?.getAttribute(`class`)),
    ).toBe(`logo`)

    expect(
      await page
        .$(`.header .target`)
        .then(async handle => await handle?.innerHTML()),
    ).toBeUndefined()
  })

  it(`should rebuild on change`, async () => {
    await update(
      path(`src`, `components`, `App.js`),
      `\
      import logo from './logo.svg'

      export const App = () => {
        return (
          <div className="App">
            <div className="header">
              <img src={logo} className="logo" alt="logo" />
              <div className="target">Noice.</div>
            </div>
          </div>
        )
      }`,
    )

    expect(
      await page
        .$(`.header img`)
        .then(async handle => await handle?.getAttribute(`class`)),
    ).toBe(`logo`)

    expect(
      await page
        .$(`.header .target`)
        .then(async handle => await handle?.innerHTML()),
    ).toBe(`Noice.`)
  })
})
