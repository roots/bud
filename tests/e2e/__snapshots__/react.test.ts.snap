// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html output of examples/react > should have expected default state 1`] = `
"import logo from './logo.svg'

export const App = () => {
  return (
    <div className="App">
      <div className="header">
        <img src={logo} className="logo" alt="logo" />
        Edit <code>src/components/App.js</code> and save to reload
      </div>
    </div>
  )
}
"
`;
